src/libs/elektra/key.c

```diff
@@ -362,32 +362,54 @@ int f (const Key * source)
  * @code
 int g (const Key * source, KeySet * ks)
 {
-	Key * dup = keyDup (source);
+	Key * dup = keyDup (source, 0);
 	// work with duplicate
 	ksAppendKey (ks, dup);
 	// ksDel(ks) will also free the duplicate
 	// source remains unchanged.
 }
  * @endcode
  *
- * Duplication of keys should be preferred to keyNew(),
- * because data like owner can be filled with a copy
- * of the key instead of asking the environment.
- * It can also be optimized in the checks, because the keyname
- * is known to be valid.
+ * Depending on what @p flags are given, only parts of the key will be copied.
+ * The supported #elektraKeyFlags values are:
+ *   - #KEY_NAME   - copies the name
+ *   - #KEY_VALUE  - copies the value
+ *   - #KEY_META   - copies all meta keys
+ *   - #KEY_BINARY - creates a binary key, even if @p source is not binary
+ *
+ * If #KEY_NAME is set, the returned Key will have the same name as @p source.
+ * Otherwise, the name will be `"/"`.
+ *
+ * If #KEY_VALUE is set, the returned Key will have the same value as @p source.*
+ * If #KEY_BINARY is set as well, the returned Key will be a binary key independent of the type of @p source.
+ *
+ * Without #KEY_VALUE, the returned Key will have an empty value. If #KEY_BINARY is set,
+ * the empty value is `NULL`, otherwise it is an emtpy string `""`.
+ *
+ * If #KEY_META is set, meta keys will be copied from @p source to the returned key.
+ * This function creates a deep copy of meta keys, i.e. all meta keys on @p source will be `keyDup`ed before
```

Comment:
There is no good way of creating a full deep copy of a key, including metadata. Via the `ksDeepDup` internal API it is possible, but a bit cumbersome:

```c
Key * k = keyDup (orig);
KeySet * meta = keyMeta (orig);
ksClear (keyMeta (k));
ksAppend (keyMeta(k), ksDeepDup(meta));
```

Also sidenote: the current `ksDeepDup` **does not** create a deep copy of metadata. So the new KeySet is not fully independent of the old one.

---

If I'm not mistaken, deep copies of metadata were not needed before we had `keyMeta`, because `keySetMeta` always uses `ksAppendKey`. But since now `keyMeta` exists the user could do:

```c
Key * k = keyDup (orig);
Key * m = ksLookupByName (keyMeta (k), "meta:/type", 0);
keySetString (m, "long");
```

Now of course `keyString (keyGetMeta (k, "type"))` is `"long"`. But if I'm not mistaken, the fact that we used a shallow copy of the metakeys means `orig` was also modified and `keyString (keyGetMeta (orig, "type"))` is `"long"`.

This is even worse, if `orig` has `KEY_RO_META` set.

Reply:
> the current ksDeepDup does not create a deep copy of metadata. So the new KeySet is not fully independent of the old one.

Can you please create an issue about it?

>  But since now keyMeta exists the user could do

Good hint, the docu needs to be updated: "You are not allowed to modify the keys or delete them."

Reply:
I created #3609

> Good hint, the docu of keyMeta needs to be updated: "You are not allowed to modify the keys or delete them."

That is a frankly stupid idea and will only lead to bugs. Instead we should add a flag to `struct _KeySet` that indicates that the whole key set is readonly. This flag will be set on `key->meta` whenever `KEY_FLAG_RO_META` is set.
With this new flag and  #3598 and #3610, there shouldn't be any problems with modifying a `keyMeta()` KeySet.

Storage plugins for example could profit a lot from operating directly on the meta KeySet. If you always just insert keys into keysets, the code should become simpler/more uniform.

Reply:
> Also sidenote: the current ksDeepDup does not create a deep copy of metadata. So the new KeySet is not fully independent of the old one.

I also encountered this "problem" when I was working on `mmapstorage`. I think at the time @markus2330 noted this behaviour of `ksDeepDup` was intentional, but I do not remember the reason (maybe performance?).

Reply:
Discussion about that is now in #3610 

> Storage plugins for example could profit a lot from operating directly on the meta KeySet. If you always just insert keys into keysets, the code should become simpler/more uniform.

Yes, exactly, this was the whole idea of `keyMeta`.

> noted this behaviour of ksDeepDup was intentional, but I do not remember the reason (maybe performance?).

I think we are talking about different uses of ksDeepDup here. When we talked about it, it was about the duplication of the KeySet during `kdbSet`.

Reply:
> the duplication of the KeySet during `kdbSet`

AFAIK this is currently the only use of `ksDeepDup`.

In any case, I now updated #3610 and we will use Copy-On-Write metadata to avoid the need for deep-copies of metadata.